### ISSUE TITLE
Query::formatResults() type hint errors

### ISSUE THREAD

**Issue by berarma** [cakephp/cakephp#10236]
This is a (multiple allowed):

* [ ] bug
* [x] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 3.3.13

I have model A that belongsTo B that belongsTo C. Model C has a behavior with a beforeFind callback. This is the beforeFind method:

```php
    public function beforeFind(Event $event, Query $query, \ArrayObject $options, $primary)
    {
        $query->formatResults(function (\Cake\Datasource\ResultSetInterface $results) {
...
        });
    }
```

When I query A.B.C the formatResults method displays error:
```Warning (4096): Argument 1 passed to App\Model\Behavior\StringPoolBehavior::App\Model\Behavior\{closure}() must implement interface Cake\Datasource\ResultSetInterface, instance of Cake\Collection\Collection given, called in vendor/cakephp/cakephp/src/ORM/Association.php on line 837 and defined [APP/Model/Behavior/StringPoolBehavior.php, line 98]```

The fastest solution would be removing type hinting. In other cases we can try to guess the type that will be used, but not this time.

Since we're most of the time expecting a ```CollectionInterface``` couldn't it be solved so we always get an object that implements this interface?

I think ```ResultSet``` should implement ```CollectionInterface```, I think it already does through a trait but it isn't declared with an ```implements```. Then ```formatResults``` would already receive a ```CollectionInterface``` in most (maybe all) cases. With this change we could safely use type hinting to make sure we get an object that implements ```CollectionInterface``` independent of the context of the call.

This way issues like mine or #10064 wouldn't happen. I think ```ResultSet``` should declare ```implements CollectionInterface``` anyway if I'm not mistaken that it does.

The documentation should be changed to use ```CollectionInterface``` wherever the ```formatResults``` method signature is shown.

What do you think?

EDIT: Changed ```ResultSetInterface``` references to ```ResultSet```.

**Comment by lorenzo** [comment]
Would you mind sending a pull request to change the docs?

**Comment by berarma** [comment]
Done, cakephp/docs#4719.

Now I think ```ResultSet``` should be changed to declare that it implements ```CollectionInterface```. I mistakenly wrote ```ResultSetInterface``` instead of ```ResultSet``` in the first post.

**Comment by berarma** [comment]
Are you waiting on me to do a new PR?

**Comment by lorenzo** [comment]
No, it is fine... I will close this now
